plugin/session_buffer.py

```diff
@@ -371,5 +390,132 @@ def _present_diagnostics_async(
         for sv in self.session_views:
             sv.present_diagnostics_async()
 
+    def do_semantic_tokens_async(self) -> None:
+        if not userprefs().semantic_highlighting:
+            return
+        if self.semantic_tokens.pending_response:
+            return
```

Comment:
I saw cases where semantic tokens weren't request after changing the document.

```
:: --> LSP-typescript textDocument/semanticTokens/full(3377): {'textDocument': {'uri': 'file:///Users/rafal/configuration/alfred/Alfred.alfredpreferences/workflows/user.workflow.81F6B074-F56F-496C-BE12-3756E6898E2A/index.js'}}
:: <<< LSP-typescript 3377: {'data': [2, 13, 7, 10, 0, 1, 6, 2, 3, 0, 0, 5, 7, 10, 0, 1, 6, 6, 7, 9, 0, 9, 7, 10, 0, 1, 20, 7, 10, 0, 1, 13, 7, 10, 0, 2, 6, 10, 7, 9, 0, 17, 6, 0, 16, 0, 7, 2, 3, 0, 0, 3, 7, 10, 0, 2, 15, 4, 10, 5, 1, 10, 9, 7, 41, 0, 19, 6, 10, 8, 0, 9, 3, 9, 1, 2, 9, 6, 11, 16, 0, 7, 4, 6, 1, 0, 8, 4, 6, 0, 0, 5, 4, 9, 8, 0, 19, 4, 6, 0, 0, 5, 3, 9, 8, 0, 8, 7, 7, 0, 0, 8, 3, 9, 0, 1, 10, 6, 7, 41, 1, 8, 5, 7, 33, 3, 8, 6, 7, 40, 1, 14, 10, 7, 41, 0, 13, 6, 7, 40, 2, 14, 12, 7, 41, 1, 14, 12, 7, 41, 0, 19, 3, 0, 16, 1, 19, 5, 7, 41, 0, 9, 12, 7, 40, 1, 18, 4, 7, 41, 0, 7, 6, 0, 16, 0, 7, 5, 7, 40, 2, 17, 4, 7, 40, 0, 5, 8, 11, 16, 0, 9, 10, 7, 40, 5, 12, 12, 7, 40, 0, 13, 3, 11, 16, 0, 4, 5, 7, 40, 0, 10, 6, 0, 16, 0, 7, 5, 7, 40, 3, 8, 7, 7, 16, 0, 8, 3, 11, 16, 0, 4, 12, 7, 40, 1, 8, 7, 7, 16, 0, 8, 3, 11, 16, 0, 4, 9, 7, 40, 2, 8, 5, 7, 32, 0, 8, 9, 7, 40, 1, 13, 6, 11, 16, 0, 7, 4, 6, 1, 0, 8, 12, 7, 40, 0, 13, 3, 11, 16, 0, 4, 4, 6, 0, 0, 5, 3, 9, 8, 1, 13, 3, 11, 16, 0, 4, 4, 6, 1, 0, 8, 6, 0, 16, 0, 7, 6, 11, 16, 0, 11, 4, 6, 0, 0, 8, 4, 9, 1, 0, 6, 12, 7, 40, 0, 13, 3, 11, 16, 0, 4, 4, 6, 0, 0, 5, 3, 9, 8, 2, 8, 5, 7, 32, 0, 26, 9, 7, 40, 3, 4, 5, 7, 32, 0, 8, 5, 7, 32, 1, 17, 4, 6, 1, 1, 22, 11, 7, 41, 0, 14, 4, 6, 0, 0, 57, 10, 7, 8, 4, 22, 11, 7, 41, 0, 14, 11, 7, 40, 2, 20, 8, 7, 33, 0, 15, 4, 6, 0, 0, 13, 11, 7, 40, 2, 20, 4, 6, 0, 2, 20, 8, 7, 32, 0, 14, 4, 6, 0, 0, 15, 8, 7, 32, 4, 20, 5, 9, 1, 0, 7, 4, 6, 0, 1, 20, 12, 9, 1, 0, 14, 4, 6, 0, 1, 20, 8, 9, 1, 1, 20, 3, 9, 1, 0, 5, 4, 6, 0, 1, 20, 4, 9, 1, 1, 24, 4, 9, 1, 1, 24, 4, 9, 1, 0, 6, 11, 7, 40, 2, 20, 4, 9, 1, 1, 24, 5, 9, 1, 1, 28, 8, 9, 1, 0, 17, 4, 6, 0, 0, 22, 4, 6, 0, 7, 16, 5, 7, 32, 3, 0, 4, 10, 4], 'resultId': '7271'}
::  -> LSP-typescript textDocument/didChange: {'contentChanges': [{'range': {'start': {'line': 34, 'character': 0}, 'end': {'line': 35, 'character': 0}}, 'rangeLength': 34, 'text': ''}, {'range': {'start': {'line': 34, 'character': 0}, 'end': {'line': 35, 'character': 0}}, 'rangeLength': 31, 'text': ''}], 'textDocument': {'version': 1622, 'uri': 'file:///Users/rafal/configuration/alfred/Alfred.alfredpreferences/workflows/user.workflow.81F6B074-F56F-496C-BE12-3756E6898E2A/index.js'}}
```

There was no `textDocument/semanticTokens/full` request after that, likely due to this check. We should instead cancel/ignore pending request and trigger a new request anyway.

Reply:
I agree, I just didn't know how to do it and couldn't find a method to cancel requests in the codebase. So I added this check as a workaround to prevent race conditions from multiple responses, because the evaluation of the response may take a while and it writes to instance variables of the SessionBuffer. I guess to cancel a request, I have to store its request id, and we alos want to make sure that a possible response is properly ignored in case the server still sends it, even though the request has been canceled. Not sure how to do that or if we already have something for this in the codebase. And maybe the request canceling & resend should better be implemented in `_when_selection_remains_stable_async` in general for all request types that are triggered on text changes and use it?

Reply:
There is a `$/cancelRequest` notification in the spec but it's not guaranteed to be supported by all servers.

> So I added this check as a workaround to prevent race conditions from multiple responses, because the evaluation of the response may take a while and it writes to instance variables of the SessionBuffer.

That shouldn't be an issue. I can see that all your code runs on the async thread so the execution of responses happens in serial.

Though I can see that you do some view offset/region calculations and change_count checks from `_draw_semantic_tokens_async` which is potentially problematic and can lead to races itself. That could should likely be run for normal thread.

Reply:
> It should be OK to ignore the response since the response can only become invalid due to document change (?) and that guarantees that a new request is already dispatched.

Though I don't know how that would play with delta updates. I suppose that in that case we always need to apply the full semantic tokens before delta can be applied. So we can't entirely ignore it.

Reply:
Yes, I already have this check for the change count to prevent drawing the highlighting regions, but the token evaluation and update will always be done. So I guess I can just remove the `pending_response` variable and corresponding check which you marked here?

Reply:
I would guess. Go ahead and do it. I can try to have a second look later when I have more time.

Reply:
Actually it's not enough to remove that check. It seems like we don't actually, re-request semantic tokens which causes this:

https://user-images.githubusercontent.com/153197/142029097-a8c25afc-9aa3-4669-8aaf-c0ddeafd02cc.mov

Reply:
My guess would be that this is the same problem we had when implementing inlay hints:

https://github.com/sublimelsp/LSP-typescript/pull/91#discussion_r725840014

Reply:
I assume this example is the race condition due to drawing from the async thread you mentioned. The buffer was probably modified right before drawing the regions, but after the change count check was performed. So I guess we must do the check + drawing from the UI thread.

Reply:
However, there should still be a new request after the buffer modification. I also saw a bug before with the rust-analyzer, where certain regions were not cleared if you comment out a line, despite the tokens have been evaluated and shown correctly via lsp_show_scope_name. This might be the same bug. I'll need to investigate further to find out what the issue is / why no new request was made.

Reply:
Currently LSP calls `self._when_selection_remains_stable_async` when it requests `self._do_semantic_tokens_async`. 

LSP currently relies on the Region being stable. (Which I think is the cause of the bug)

But I would suggest switching to `view.change_count` just to see if it will fix the problem. 

You can see this commit if you are interested in such fix https://github.com/sublimelsp/LSP-typescript/commit/8e0d46890fd5b78824703125acbfa51ff074ce41

Reply:
Yes, relying on `_when_selection_remains_stable_async` doesn't seem right since that can result in not making a request when just moving cursor quickly after changing a document.

I suppose other features might suffer from the same issue.

Reply:
I assumed this would already be handled by `_when_selection_remains_stable_async`, i.e. the request might be delayed until the selection doesn't change anymore, but it is still guaranteed that the request is always made (after some time).

So we need a `_when_buffer_content_remains_stable_async` method for the requests triggered by the text change listener?

Reply:
1. Another issue is that the tokens need to be requested after the `didChange` notification (sent from `SessionBuffer.purge_changes_async`). That's most likely what causes my issue.

2. You are currently requesting from `DocumentSyncListener` which could have multiple sessions. You could argue that it only makes sense to request from one of the sessions that support semantic tokens but there could be a case when it makes sense to have tokens per session. For example for files like Vue files that host JS/CSS/HTML each session might apply token to the block that it handles.

Reply:
For the debouncing issue I think you could create an instance of `Debouncer` and then call

```py
self._debouncer_semantic_tokens.debounce(self._do_semantic_tokens_async,
                                                 timeout_ms=self.semantic_tokens_debounce_time,
                                                 async_thread=True)
```

Reply:
I've created https://github.com/sublimelsp/LSP/pull/1899 to get this sorted for color boxes.
Once we agree that the solution is fine, you can probably copy that approach here.
(We also want to fix code lenses the same way I guess)

Reply:
This works well now.

Reply:
I've noticed another bug due to the removal of this "pending response" check. If we send requests for semantic tokens delta, and it happens that a 2nd request is sent before the response for the first request arrived, then it can happen that our data for semantic tokens gets out of sync with the server. Because both deltas were relative to the same "resultId" (of the previous response), but we would just apply them to our latest version.

So if the communication is like this:

```
--> request(1) / params: "previousResultId" = 42
--> request(2) / params: "previousResultId" = 42
<-- response(1) / params: "resultId" = 43
<-- response(2) / params: "resultId" = 44
```

I have seen this with rust-analyzer.

For the delat request, the server can either answer with a SemanticTokensDelta type for the response (difference to previous result), or with a SemanticTokens type (full token data).

Fortunately, rust-analyzer everytime just sends the full token data with the response (type: SemanticTokens), so all the raw token data is just overwritten on our side, and this potential problem never happens with rust-analyzer. (I don't know why they would advertise support for "semanticTokensProvider.full.delta" when they never actually use it...)

But in general this can still be a problem. So I either need to reintroduce this "pending_response" variable and prevent requests if the former response has not arrived (this check could be moved only into the branch where we send Request.semanticTokensFullDelta). This would mean that sometimes no request is made even when there were additional changes in the buffer, and semantic highlighting might look incomplete until there is a new change of the buffer content + request.
Or we need to keep multiple versions of our raw semantic tokens data around, with the corresponding "resultId"s, and then also keep track of the request ids and associate them to the token "resultId"s somehow.